### PR TITLE
Time Complexity

### PR DESCRIPTION
**Pull request by HarryWangATX**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: content/2_Bronze/Time_Comp.mdx [hunk 1]
```diff
@@ -440,19 +433,46 @@ addition operations take a bit longer than a single addition operation. Another
...
 
-<!-- For example, if our code currently runs in $\mathcal{O}(n^2)$ time, perhaps we can modify our code to make it run in $\mathcal{O}(n^2/32)$ by using a bitset. (Of course, with big-O notation, $\mathcal{O}(n^2) = O(n^2/32)$.) -->
+<!-- For example, if our code currently runs in $\mathcal{O}(n^2)$ time, perhaps we can modify our code to make it run in $\mathcal{O}(n^2/32)$ by using a bitset. (Of course, with Big O notation, $\mathcal{O}(n^2) = O(n^2/32)$.) -->
 <!-- Bitsets are significantly faster (and so are bitwise operations vs iterating through bits). I don't want to make it sound insignificant -->
 
 For now, don't worry about optimizing constant factors -- just be aware of them.
 
+## Formal Definition of Big O notation
+
+Let $f$ and $g$ be non-negative functions from $\mathbb{R}_{\ge 0}$ to $\mathbb{R}_{\ge 0}$. If there exists
```
Comment (bqi343) [inline content/2_Bronze/Time_Comp.mdx:451]:
```suggestion
Let $f$ and $g$ be non-negative functions from $\mathbb{R}_{\ge 0}$ to $\mathbb{R}_{\ge 0}$. If there exist

```

File: content/2_Bronze/Time_Comp.mdx [hunk 2]
```diff
@@ -440,19 +433,46 @@ addition operations take a bit longer than a single addition operation. Another
...
+
+Therefore, it is correct to say that the time complexity of a linear function, $\mathcal{O}(n)$, is $\mathcal{O}(n^2)$, $\mathcal{O}(2^n)$, $\mathcal{O}(n^n)$, etc.
+
+But typically, we write the most restrictive function, thus, $\mathcal{O}(n)$.
+
+<Optional title="P vs. NP">
+
+P refers to the class of problems that can be solved within polynomial time. (ie. $\mathcal{O}(n^2)$, $\mathcal{O}(n^3)$, $\mathcal{O}(n^100)$...)
+NP, known as nondeterministic polynomial, is the set of problems with solutions that can be verified in polynomial time, but not 
+solved in polynomial time.
```
Comment (bqi343) [inline content/2_Bronze/Time_Comp.mdx:463]: it seems like you're asserting that P != NP here

File: content/2_Bronze/Time_Comp.mdx [hunk 3]
```diff
@@ -440,19 +433,46 @@ addition operations take a bit longer than a single addition operation. Another
...
+
+A common example of NP is the game of Sudoku, where a solution is easily verifiable, but not easily computable (exponentially slower as grid size increases).
+
+“P = NP” is the classic problem that determines whether a problem that can be verified in polynomial time can also be solved
```
Comment (bqi343) [inline content/2_Bronze/Time_Comp.mdx:467]:
```suggestion
“P vs NP” is the classic problem that asks whether every problem that can be verified in polynomial time can also be solved 

```

Comment (bqi343) [reply content/2_Bronze/Time_Comp.mdx:467]: also mention that this is unsolved

File: content/2_Bronze/Time_Comp.mdx [hunk 4]
```diff
@@ -440,19 +433,46 @@ addition operations take a bit longer than a single addition operation. Another
...
+solved in polynomial time.
+
+A common example of NP is the game of Sudoku, where a solution is easily verifiable, but not easily computable (exponentially slower as grid size increases).
+
+“P = NP” is the classic problem that determines whether a problem that can be verified in polynomial time can also be solved 
+in polynomial time.
+
+If you're interested in learning more about P vs. NP, check out this
+[Youtube video](https://www.youtube.com/watch?v=YX40hbAHx3s). If you're just
+starting out, you may want to skip this video for now.
```
Comment (bqi343) [inline content/2_Bronze/Time_Comp.mdx:472]:
```suggestion
If you're interested in learning more about P vs. NP, check out this
[Youtube video](https://www.youtube.com/watch?v=YX40hbAHx3s).

```

seems unnecessary to include this line, considering that everything is inside an optional block

File: content/2_Bronze/Time_Comp.mdx [hunk 5]
```diff
@@ -440,19 +433,46 @@ addition operations take a bit longer than a single addition operation. Another
...
 
 For now, don't worry about optimizing constant factors -- just be aware of them.
 
+## Formal Definition of Big O notation
+
+Let $f$ and $g$ be non-negative functions from $\mathbb{R}_{\ge 0}$ to $\mathbb{R}_{\ge 0}$. If there exists
+positive constants $n_0$ and $c$ such that $f(n) \le c \cdot g(n)$ whenever $n \ge n_0$, we
+say that $f(n) = \mathcal{O}(g(n))$.
+
+Therefore, it is correct to say that the time complexity of a linear function, $\mathcal{O}(n)$, is $\mathcal{O}(n^2)$, $\mathcal{O}(2^n)$, $\mathcal{O}(n^n)$, etc.
```
Comment (bqi343) [inline content/2_Bronze/Time_Comp.mdx:455]:
```suggestion
Therefore, it is correct to say that the time complexity of a linear function, $\mathcal{O}(n)$, is $\mathcal{O}(n/2)$, $\mathcal{O}(2n)$, $\mathcal{O}(n^2)$, $\mathcal{O}(2^n)$, $\mathcal{O}(n^n)$, etc.

```

File: content/2_Bronze/Time_Comp.mdx [hunk 6]
```diff
@@ -440,19 +433,46 @@ addition operations take a bit longer than a single addition operation. Another
...
+
+But typically, we write the most restrictive function, thus, $\mathcal{O}(n)$.
```
Comment (bqi343) [inline content/2_Bronze/Time_Comp.mdx:457]:
```suggestion
But typically, we write the simplest function out of those that are the most restrictive, thus, $\mathcal{O}(n)$.

```

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/2_Bronze/Time_Comp.mdx [hunk 7]
```diff
@@ -440,19 +433,50 @@ addition operations take a bit longer than a single addition operation. Another
...
+
+P refers to the class of problems that can be solved within polynomial time. (ie. $\mathcal{O}(n^2)$, $\mathcal{O}(n^3)$, $\mathcal{O}(n^100)$...)
+NP, known as nondeterministic polynomial, is the set of problems with solutions that can be verified in polynomial time.
+
+A common example of NP is the game of Sudoku, where a solution is easily verifiable, but not easily computable (exponentially slower as grid size increases), though 
+there may exist a solution that does solve the solution under polynomial time.
+
+“P vs NP” is the classic problem that asks whether every problem that can be verified in polynomial time can also be solved. This question is currently unsolved. 
+
+in polynomial time.
```
Comment (bqi343) [inline content/2_Bronze/Time_Comp.mdx:472]: ^ move this to the correct place

Comment (bqi343) [reply content/2_Bronze/Time_Comp.mdx:472]: I'll fix it

**Review by bqi343 — APPROVED** [review]

File: content/2_Bronze/Time_Comp.mdx [hunk 8]
```diff
@@ -440,19 +433,45 @@ addition operations take a bit longer than a single addition operation. Another
...
+
+But typically, we write the simplest function out of those that are the most restrictive, thus, $\mathcal{O}(n)$.
+
+
+<Optional title="P vs. NP">
+
+P refers to the class of problems that can be solved within polynomial time ($\mathcal{O}(n^2), \mathcal{O}(n^3), \mathcal{O}(n^{100}), \dots$).
+NP, short for nondeterministic polynomial time, is the set of problems with solutions that can be verified in polynomial time.
+
+A common example of a problem in NP is a generalized version of Sudoku, where a solution is easily verifiable in polynomial time, but it is unknown whether a solution is computable in polynomial time. “P vs. NP” is the classic unsolved problem that asks whether every problem that can be verified in polynomial time can also be solved in polynomial time.
```
Comment (SansPapyrus683) [inline content/2_Bronze/Time_Comp.mdx:467]: > “P vs. NP” is the classic

quotes seem a bit different?

File: content/2_Bronze/Time_Comp.mdx [hunk 9]
```diff
@@ -440,19 +433,45 @@ addition operations take a bit longer than a single addition operation. Another
...
 For now, don't worry about optimizing constant factors -- just be aware of them.
 
+## Formal Definition of Big O notation
+
+Let $f$ and $g$ be non-negative functions from $\mathbb{R}_{\ge 0}$ to $\mathbb{R}_{\ge 0}$. If there exist
+
+positive constants $n_0$ and $c$ such that $f(n) \le c \cdot g(n)$ whenever $n \ge n_0$, we
+say that $f(n) = \mathcal{O}(g(n))$.
+
+Therefore, it is correct to say that the time complexity of a linear function, $\mathcal{O}(n)$, is $\mathcal{O}(n/2)$, $\mathcal{O}(2n)$, $\mathcal{O}(n^2)$, $\mathcal{O}(2^n)$, $\mathcal{O}(n^n)$, etc.
```
Comment (SansPapyrus683) [inline content/2_Bronze/Time_Comp.mdx:456]:
```suggestion
Therefore, we *could* say that the time complexity of a linear function, $\mathcal{O}(n)$, is also $\mathcal{O}(n/2)$, $\mathcal{O}(2n)$, $\mathcal{O}(n^2)$, $\mathcal{O}(2^n)$, $\mathcal{O}(n^n)$, etc.
However, we usually just write the simplest function out of those that are the most restrictive, which in the case of our linear function above, is $\mathcal{O}(n)$.

File: content/2_Bronze/Time_Comp.mdx [hunk 10]
```diff
@@ -440,19 +433,45 @@ addition operations take a bit longer than a single addition operation. Another
...
+<!-- For example, if our code currently runs in $\mathcal{O}(n^2)$ time, perhaps we can modify our code to make it run in $\mathcal{O}(n^2/32)$ by using a bitset. (Of course, with Big O notation, $\mathcal{O}(n^2) = O(n^2/32)$.) -->
 <!-- Bitsets are significantly faster (and so are bitwise operations vs iterating through bits). I don't want to make it sound insignificant -->
 
 For now, don't worry about optimizing constant factors -- just be aware of them.
 
+## Formal Definition of Big O notation
+
+Let $f$ and $g$ be non-negative functions from $\mathbb{R}_{\ge 0}$ to $\mathbb{R}_{\ge 0}$. If there exist
+
+positive constants $n_0$ and $c$ such that $f(n) \le c \cdot g(n)$ whenever $n \ge n_0$, we
```
Comment (SansPapyrus683) [inline content/2_Bronze/Time_Comp.mdx:453]:
```suggestion
Let $f$ and $g$ be non-negative functions from $\mathbb{R}_{\ge 0}$ to $\mathbb{R}_{\ge 0}$. If there exist
positive constants $n_0$ and $c$ such that $f(n) \le c \cdot g(n)$ whenever $n \ge n_0$, we
```

File: content/2_Bronze/Time_Comp.mdx [hunk 11]
```diff
@@ -440,19 +433,45 @@ addition operations take a bit longer than a single addition operation. Another
...
+say that $f(n) = \mathcal{O}(g(n))$.
+
+Therefore, it is correct to say that the time complexity of a linear function, $\mathcal{O}(n)$, is $\mathcal{O}(n/2)$, $\mathcal{O}(2n)$, $\mathcal{O}(n^2)$, $\mathcal{O}(2^n)$, $\mathcal{O}(n^n)$, etc.
+
+
+But typically, we write the simplest function out of those that are the most restrictive, thus, $\mathcal{O}(n)$.
+
```
Comment (SansPapyrus683) [inline content/2_Bronze/Time_Comp.mdx:460]:
```suggestion
```

**Review by SansPapyrus683 — COMMENTED** [review]
image

File: content/2_Bronze/Time_Comp.mdx (hunk 11 above)
Comment (HarryWangATX) [reply content/2_Bronze/Time_Comp.mdx:459]: i think this is neede cuz it defines what restrictive and simple means, especially on a beginner lvl.

File: content/2_Bronze/Time_Comp.mdx (hunk 8 above)
Comment (HarryWangATX) [reply content/2_Bronze/Time_Comp.mdx:466]: not sure i understand what u mean

File: content/2_Bronze/Time_Comp.mdx (hunk 11 above)
Comment (SansPapyrus683) [reply content/2_Bronze/Time_Comp.mdx:460]: i included it in the previous suggestion

File: content/2_Bronze/Time_Comp.mdx (hunk 8 above)
Comment (SansPapyrus683) [reply content/2_Bronze/Time_Comp.mdx:467]: `”` vs `"`

**Pull request merged**